File: src/main/java/su/nlq/idea/externalizable/modification/CleanupMethods.java
package su.nlq.idea.externalizable.modification;

import com.intellij.openapi.project.Project;
import com.intellij.psi.*;
import org.jetbrains.annotations.NotNull;

import java.util.function.Predicate;
import java.util.stream.Stream;

public final class CleanupMethods implements PsiClassModification {

  @Override
  public void accept(@NotNull PsiClass psiClass) {
    final Project project = psiClass.getProject();
    Stream.of(psiClass.getMethods())
        .filter(
            new MethodNameFilter("readExternal").and(new SingleParameterTypeFilter(project, "java.io.ObjectInput"))
                .or(new MethodNameFilter("writeExternal").and(new SingleParameterTypeFilter(project, "java.io.ObjectOutput"))))
        .forEach(PsiElement::delete);
  }

  private static final class MethodNameFilter implements Predicate<PsiMethod> {
    @NotNull
    private final String name;

    public MethodNameFilter(@NotNull String name) {
      this.name = name;
    }

    @Override
    public boolean test(@NotNull PsiMethod method) {
      return method.getName().equals(name);
    }
  }

  private static final class SingleParameterTypeFilter implements Predicate<PsiMethod> {
    @NotNull
    private final PsiClassType type;

    public SingleParameterTypeFilter(@NotNull Project project, @NotNull String type) {
      this.type = new TypeFinder(project, type).get();
    }

    @Override
    public boolean test(@NotNull PsiMethod method) {
      final PsiParameter[] parameters = method.getParameterList().getParameters();
      return parameters.length == 1 && parameters[0].getType().equals(type);
    }
  }
}

File: src/main/kotlin/su/nlq/idea/externalizable/modification/CleanupMethods.kt
package su.nlq.idea.externalizable.modification

import com.intellij.openapi.project.Project
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiClassType
import com.intellij.psi.PsiMethod
import java.util.function.Predicate
import java.util.stream.Stream

class CleanupMethods : PsiClassModification {

    override fun modify(psiClass: PsiClass) {
        val project = psiClass.project
        Stream.of(*psiClass.methods)
            .filter(
                MethodNameFilter("readExternal")
                    .and(SingleParameterTypeFilter(project, "java.io.ObjectInput"))
                    .or(
                        MethodNameFilter("writeExternal")
                            .and(SingleParameterTypeFilter(project, "java.io.ObjectOutput"))
                    )
            )
            .forEach { it.delete() }
    }

    private class MethodNameFilter(private val name: String) : Predicate<PsiMethod> {

        override fun test(method: PsiMethod) = method.name == name
    }

    private class SingleParameterTypeFilter(project: Project, type: String) : Predicate<PsiMethod> {
        private val type: PsiClassType = TypeFinder(project, type).get()

        override fun test(method: PsiMethod): Boolean {
            val parameters = method.parameterList.parameters
            return parameters.size == 1 && parameters[0].type == type
        }
    }
}
